### ISSUE TITLE
Encode buf.yaml in Module

### ISSUE THREAD

**Issue by amckinney** [bufbuild/buf#682]
We currently encode `dependencies` (i.e. the `buf.lock`) and `documentation` (i.e. the `buf.md`) in the `Module`, but we do not encode anything found in the `buf.yaml`. 

With the `buf.yaml`, the BSR could start to enforce `lint` and `breaking` configurations on the server-side, and we could more easily include license information which we currently plan to include in the `buf.yaml` (re: https://github.com/bufbuild/buf/issues/488).

We'll need to start by modeling the contents of the `buf.yaml` in `.proto`, which will roughly look something like the following:

```diff
diff --git a/proto/buf/alpha/module/v1alpha1/module.proto b/proto/buf/alpha/module/v1alpha1/module.proto
index 5b79ac80..8e4a772a 100644
--- a/proto/buf/alpha/module/v1alpha1/module.proto
+++ b/proto/buf/alpha/module/v1alpha1/module.proto
@@ -34,6 +34,14 @@ message Module {
   //
   // string is used to enforce UTF-8 encoding or 7-bit ASCII text.
   string documentation = 3;
+  // license is the string representation of the license.
+  string license = 4;
+  // module_name is the name.
+  ModuleName module_name = 5;
+  // lint_config is the lint configuration that can (optionally) be enforced by the BSR.
+  LintConfig lint_config = 6;
+  // breaking_config is the breaking configuration that can (optionally) be enforced by the BSR.
+  BreakingConfig breaking_config = 7;
 }
```

Note that we _don't_ need to include the [build](https://docs.buf.build/configuration/v1/buf-yaml#build) configuration found in the `buf.yaml` since this might be deprecated in favor of `.bufignore` (re: https://github.com/bufbuild/buf/issues/578).

**Comment by doriable** [comment]
Some things we discussed today regarding the encoding the module configuration for use on the server side:

- While it is a "nice side effect" to be able to "reconstruct" the module configuration for the user through the UI, this should be considered a non-goal for this. The primary requirement is to be able to use module configuration for server-side work, such as gating pushes with lint and breaking change detection, etc. We can eventually surface the module configuration to the user, but it won't be the exact representation of their `buf.yaml` (e.g. comments in the `buf.yaml`, whitespace, etc. are not included) since we are not working with the raw representation of the `buf.yaml`
- We have decided, as the issue description has also noted, that we don't need to include the build configuration. This is because the module representation that we're receiving on the server side would already have the build configurations applied (e.g. `build.excludes` paths would already be excluded) and is not needed on the server side for working with the module. This could change in the future, but for now, we won't be including them. This also factors into the previous point, where we will not be able to show users an exact representation of their `buf.yaml` without the build configurations, for example.
- Lastly, configuration versions should be supported for lint and breaking change detection configs. Right now, the shape of the `v1beta1` and `v1` configuration shapes for breaking and lint are the same shape, but we should explicitly send through the version of the configuration we're supporting.

**Comment by doriable** [comment]
Re-opening this for porting to the BSR.

**Comment by amckinney** [comment]
As I was working on fixing #715, I encountered an issue with #739 - we're not preserving the 1:1 property between the module's Protobuf representation and its internal `bufmodule.Module` representation.

The mapping from the Protobuf representation to the `bufmodule.Module` representation is lossy - a variety of `lint` configuration fields are not persisted in the [buflintconfig.Config](https://github.com/bufbuild/buf/blob/eeed36e27cddefb164f09eba8e6e8a326047a37e/private/buf/bufcheck/buflint/buflintconfig/buflintconfig.go#L39) type - they are properties of the [internal.ConfigBuilder](https://github.com/bufbuild/buf/blob/eeed36e27cddefb164f09eba8e6e8a326047a37e/private/buf/bufcheck/internal/config.go#L47).

This means that once we transform into the `bufmodule.Module` type, it's currently not possible for us to recover the original `lint` configuration set by the user. When the user builds the module and pushes it, the Protobuf representation wouldn't actually contain these `lint` values, and the digest wouldn't reflect it either.

We need to revisit this implementation and propagate these `lint` configuration values through the `bufmodule.Module` type so that we maintain the 1:1 relationship. I've reverted #739 for now in [this](https://github.com/bufbuild/buf/commit/eeed36e27cddefb164f09eba8e6e8a326047a37e) commit so that we aren't in an interim unreleasable state, but _most_ of the functionality will be reintroduced in a follow-up.
